crates/blockchain/blockchain.rs

```diff
@@ -74,45 +84,67 @@ impl Blockchain {
         );
         let execution_result = vm.execute_block(block)?;
 
+        // Validate execution went alright
+        validate_gas_used(&execution_result.receipts, &block.header)?;
+        validate_receipts_root(&block.header, &execution_result.receipts)?;
+        validate_requests_hash(&block.header, &chain_config, &execution_result.requests)?;
+
         Ok(execution_result)
     }
 
-    pub fn store_block(
+    /// Executes a block from a given vm instance an does not clear its state
+    fn execute_block_from_state(
         &self,
+        parent_header: &BlockHeader,
         block: &Block,
-        execution_result: BlockExecutionResult,
-    ) -> Result<(), ChainError> {
-        // Assumes block is valid
-        let BlockExecutionResult {
-            receipts,
-            requests,
-            account_updates,
-        } = execution_result;
-        let chain_config = self.storage.get_chain_config()?;
+        chain_config: &ChainConfig,
+        vm: &mut Evm,
+    ) -> Result<BlockExecutionResult, ChainError> {
+        // Validate the block pre-execution
+        validate_block(block, parent_header, chain_config)?;
 
-        let block_hash = block.header.compute_block_hash();
+        let execution_result = vm.execute_block_without_clearing_state(block)?;
 
-        validate_gas_used(&receipts, &block.header)?;
+        // Validate execution went alright
+        validate_gas_used(&execution_result.receipts, &block.header)?;
+        validate_receipts_root(&block.header, &execution_result.receipts)?;
+        validate_requests_hash(&block.header, chain_config, &execution_result.requests)?;
+
+        Ok(execution_result)
+    }
 
-        // Apply the account updates over the last block's state and compute the new state root
-        let new_state_root = self
+    pub fn store_block(
+        &self,
+        block: &Block,
+        execution_result: BlockExecutionResult,
+    ) -> Result<(), ChainError> {
+        // Apply the account updates over the block's parent state and compute the new state root
+        let Some(mut state_trie) = self
             .storage
-            .apply_account_updates(block.header.parent_hash, &account_updates)?
-            .ok_or(ChainError::ParentStateNotFound)?;
+            .state_trie(block.header.parent_hash)
+            .map_err(ChainError::StoreError)?
+        else {
+            return Err(ChainError::ParentStateNotFound);
+        };
 
+        self.storage
```

Comment:
I don't understand why we're decoupling hash tree from commit here since we're just storing one single block?